mautic/mautic
# Form object refactoring

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | /
| Related developer documentation PR URL | TODO (update the API to not use the deprecated field)
| Issues addressed (#s or URLs) | https://forum.mautic.org/t/mautic-custom-objects-plugin/24058/8?u=escopecz
| BC breaks? | N
| Deprecations? | Y

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

⚠️ This PR is using the Event contracts needed for Symfony 5 so it does not have to be refactored again for M5. https://github.com/mautic/mautic/pull/11140 must be merged first so this can be tested.

[//]: # ( Required: )
#### Description:

The final goal is that Custom Objects will be able to inject their objects and fields into the form field mapping form.

- [x] Objects and fields must be subscribed by a listener. One event for objects, the other to get all fields for a specific object.
- [x] New select box for objects
- [x] The fields select box will update its options based on selected object in the previous select box via AJAX.
- [x] Create new columns in the form_fields table: mapped_object, mapped_field
- [x] In the migration populate the new columns with the value from the lead_field column. Fill in the right object (lead, company) based on the value prefix.
- [x] Deprecate the lead_field column in the entity.
- [x] Replace usage of deprecated lead_field with mapped_field and mapped_object values.
- [x] Replace usages of 'contactFields' and 'companyFields'
- [x] Still save the value to the deprecated field for API BC compatibility.
- [x] Make sure that if a form is created/modified with the legacy field that it will populate the replacement fields for API BC compatibility.
- [x] Store field ~~IDs~~ or aliases?

Contains also migration that will create the columns and migrates data from lead_field column.

I cleaned the entity a bit too. It was full of verbose syntax, strings over constants and comments that did not bring any value.

#### Steps to test this PR:
1. Before you checkout this branch:
2. Create a form with at least 4 fields. One mapped to the contact email field, one mapped to the company field under the contact opt group, one mapped to the primary company city field. The fourth field should not be mapped to any field.
3. Run the migration. Go to the `form_fields` table and check that the last 2 columns are `mapped_object` and `mapped_field`. The field that was mapped to the contact email field should have `mapped_object = 'lead'` and `mapped_field = 'email'`. The one that was mapped to the company field under the contact opt group should have `mapped_object = 'lead'` and `mapped_field = 'company'`. The one that was mapped to the primary company city field should have `mapped_object = 'company'` and `mapped_field = 'companycity'`. The form field that was not mapped at all should have null values there.
4. When you map one field then it should be removed from mapping of other fields. Example: Map contact email field to a form field. Create another field. The contact email field is not in the mapping list.
5. Test submissions of different field types.
6. Test conditional fields
7. Test progressive profiling (behaviours)

#### List deprecations along with the new alternative:
1. `form_fields.lead_field` column is deprecated. Use `mapped_object` and `mapped_field` instead.
2. `Mautic\FormBundle\Entity\Field::getLeadField()` and `setLeadField()`. Use getters and setters on `mappedObject` and `mappedField` properties instead.
3. `Mautic\FormBundle\Event\Service\FieldValueTransformer::isIsTransformed()` is deprecated as not used.

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/11234"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
